distributed_shampoo/utils/shampoo_preconditioner_list.py

```diff
@@ -1119,13 +1134,30 @@ def _create_kronecker_factors_state_for_block(
             ),
         )
 
+    def _get_inverse_exponent_override_based(self, dimension: int, order: int) -> float:
```

Comment:
Is there any difference between this and the other definition of this function? Is it possible to move it to the base class? It could be overwritten for `EigenvalueCorrectedShampooPreconditionerList`.

Reply:
Regarding the shared intermediate class, see my comment about specifying the types for the `Generic` base class.

> Overwriting this only in EigenvalueCorrectedShampooPreconditionerList might be a technical debt in the long run because the intention of overwriting a concrete function like _get_inverse_exponent_override is conditional based on the class.

Is `_get_inverse_exponent` not a function that every child class will have to implement either way, even if implicitly like it is currently done for `EigenvalueCorrectedShampooPreconditionerList`?

Reply:
But it could also be an abstract method because any Shampoo-like algorithm (all child classes of `BaseShampooPreconditionerList`) will have a notion of an inverse exponent.

Reply:
I especially like that we can factor out `_create_preconditioned_dims_selector` and `_get_inverse_roots_from_override`. My only worry is that the inheritance structure will become too complex over time, especially with all the `TypeVars`, but let's go for this for now.